File: artifacts/tcp_client.js
const net = require('net');

const client = new net.Socket();

// Only one request to send
const request = { method: 'word', word: 'hello' };

let running = true;

client.connect(5000, '0.0.0.0', () => {
  console.log('Connected to server');
  // Send the single request
  if (running) {
    console.log('Sending request:', request);
    client.write(JSON.stringify(request));
  }
});

client.on('data', (data) => {
  console.log('Received:', data.toString());
  // Close after receiving the response to the only request
  running = false;
  client.end();
});

client.on('close', () => {
  console.log('Connection closed');
});

client.on('error', (err) => {
  console.error('Error:', err);
});

// Handle Ctrl+C to stop the client gracefully
process.on('SIGINT', () => {
  console.log('\nStopping client...');
  running = false;
  client.end();
  process.exit(0);
});

File: artifacts/tcp_client.cpp
#include "artifacts/tcp_client.h"

#include <string>

TcpClient::TcpClient() : socket(TcpSocket()) {}

TcpClient::~TcpClient() {
    socket.disconnect();
}

Result<void*> TcpClient::connect(const std::string& host, int port) {
    logger.info("Connecting to server " + host + ":" + std::to_string(port));
    return socket.connect(host, port).finally<void*>([&]() {
        return nullptr;
    });
}

Result<void*> TcpClient::disconnect() {
    logger.info("Disconnecting from server");
    return socket.disconnect();
}

Result<std::string> TcpClient::request(const std::string& data) {
    logger.info("Sending request to server: " + data);
    return socket.send(data).and_then<std::string>([&](auto) {
        return socket.receive();
    });
}